### PR TITLE
Examine error code returned by curl handle instead of return value from curl_exec. Using the return value from curl_exec resulted in failure on new mailboxes that did not contain any mail.

### LINKED ISSUE #154: IMAP backend doesn't work with empty mailbox
### Steps to reproduce
1. Configure IMAP backend
2. Create new IMAP user
3. Try to login to NC

### Expected behaviour
Login to NC

### Actual behaviour
"Wrong username or password" message is shown. But in IMAPS server log I can see a successful login!
After sending at least one e-mail to the newly created mailbox, login works as expected. After deleting e-mail, user is unable to login again.

### Affected Authentication backend
IMAP only

### Server configuration

**User External App version:**  0.10.0

**Operating system:** CentOS7

**Web server:** Apache 2.4.6

**Database:** MariaDB 5.5

**PHP version:** PHP 7.3.9

**Nextcloud version:** 18.0.7

**Updated from an older Nextcloud/ownCloud or fresh install:** fresh

**Where did you install Nextcloud from:**

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.4.0
  - activity: 2.11.0
  - calendar: 2.0.3
  - cloud_federation_api: 1.1.0
  - comments: 1.8.0
  - contacts: 3.3.0
  - dav: 1.14.0
  - federatedfilesharing: 1.8.0
  - federation: 1.8.0
  - files: 1.13.1
  - files_pdfviewer: 1.7.0
  - files_rightclick: 0.15.2
  - files_sharing: 1.10.1
  - files_trashbin: 1.8.0
  - files_versions: 1.11.0
  - files_videoplayer: 1.7.0
  - firstrunwizard: 2.7.0
  - groupfolders: 6.0.6
  - logreader: 2.3.0
  - lookup_server_connector: 1.6.0
  - mail: 1.4.1
  - nextcloud_announcements: 1.7.0
  - notifications: 2.6.0
  - oauth2: 1.6.0
  - password_policy: 1.8.0
  - photos: 1.0.0
  - privacy: 1.2.0
  - provisioning_api: 1.8.0
  - recommendations: 0.6.0
  - serverinfo: 1.8.0
  - settings: 1.0.0
  - sharebymail: 1.8.0
  - support: 1.1.1
  - survey_client: 1.6.0
  - systemtags: 1.8.0
  - text: 2.0.0
  - theming: 1.9.0
  - twofactor_backupcodes: 1.7.0
  - twofactor_nextcloud_notification: 2.3.0
  - twofactor_totp: 4.1.3
  - updatenotification: 1.8.0
  - user_external: 0.10.0
  - viewer: 1.2.0
  - workflowengine: 2.0.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "nc.example.com"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "18.0.7.1",
        "overwrite.cli.url": "http:\/\/localhost",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "default_locale": "cs_CZ",
        "logfile": "\/mnt\/data\/nextcloud\/log\/nextcloud.log",
        "log_rotate_size": 0,
        "logtimezone": "Europe\/Prague",
        "skeletondirectory": "\/var\/www\/nextcloud.local\/skeleton",
        "apps_paths": [
            {
                "path": "\/var\/www\/nextcloud\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/nextcloud\/apps.local",
                "url": "\/apps.local",
                "writable": true
            }
        ],
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "tls",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": "0",
            "dbindex": "1",
            "password": "***REMOVED SENSITIVE VALUE***",
            "timeout": "1.5"
        },
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mysql.utf8mb4": true,
        "data-fingerprint": "e57489a8e8673388fc3d8b621ddeabfe",
        "theme": "",
        "loglevel": 2,
        "updater.release.channel": "stable",
        "user_backends": [
            {
                "class": "OC_User_IMAP",
                "arguments": [
                    "imap.example.com",
                    993,
                    "ssl",
                    "example.com",
                    true,
                    true
                ]
            }
        ]
    }
}

```
</details>

#### Web server error log
<details>
<summary>Web server error log</summary>

```
No errors in web server log
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
{"reqId":"XyQeMuc7w8DuQ2wnmSdtKAAAAAo","level":3,"time":"2020-07-31T15:35:48+02:00","remoteAddr":"XXX.XXX.XXX.XXX","user":"--","app":"user_external","method":"POST","url":"/index.php/login","m
essage":"ERROR: Could not connect to imap server via curl: ","userAgent":"Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36","ve
rsion":"18.0.7.1"}
{"reqId":"XyQeMuc7w8DuQ2wnmSdtKAAAAAo","level":2,"time":"2020-07-31T15:35:48+02:00","remoteAddr":"XXX.XXX.XXX.XXX","user":"--","app":"no app in context","method":"POST","url":"/index.php/login
","message":"Login failed: john.doe@example.com (Remote IP: XXX.XXX.XXX.XXX)","userAgent":"Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.8
9 Safari/537.36","version":"18.0.7.1"}
```
</details>

#### IMAP server (Dovecot) log from the **same time**
<details>
<summary>Dovecot log</summary>

```
Jul 31 15:35:48 nc.example.com dovecot[281401]: imap-login: Login: user=<john.doe>, method=LOGIN, rip=::1, lip=::1, mpid=282462, TLS, session=<uWzj3cxrmpMAAAAAAAAAAAAAAAAAAAAB>
Jul 31 15:35:48 nc.example.com dovecot[281401]: imap(john.doe): Logged out in=46 out=831
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
No errors in browser log```
</details>

### PR DESCRIPTION
**Pull request by tem-hth**
fixes #154

### CODE REVIEW

**Review by brknkfr — APPROVED** [review]
Seems a reasonable change.

**Comment by violoncelloCH** [comment]
thanks @tem-hth 
I guess this possibly fixes #154 right?

File: lib/imap.php [hunk 1]
```diff
@@ -97,9 +97,10 @@ public function checkPassword($uid, $password) {
 		curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 10);
 		curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'CAPABILITY');
 
-		$canconnect = curl_exec($ch);
+		curl_exec($ch);
+		$errorcode = curl_errno($ch);
 
-		if($canconnect) {
+		if($errorcode == 0) {
```
Comment (violoncelloCH) [inline lib/imap.php:103]: could you use an explicit comparison  (`===`)?

**Comment by violoncelloCH** [comment]
also, there are 3 commits for those few changes... could you rebase and squash them (into a single commit) and also add the sign-off? 
(https://github.com/nextcloud/server/blob/master/.github/CONTRIBUTING.md#sign-your-work)

File: lib/imap.php (hunk 1 above)
Comment (albansch) [reply lib/imap.php:103]: FYI I've manually used this fix in one of my nextcloud 20.0.10 instance, and it has resolved the issue I was facing with IMAP users having 0 email in their mailbox not being able to login, as wrong username/password message was coming back. THX! This shall go into next release of nextcloud.

`               curl_exec($ch);
                $errorcode = curl_errno($ch);

                if($errorcode === 0) {`

Comment (tem-hth) [reply lib/imap.php]: I made the requested updates to my code.

**Comment by richarson** [comment]
Hello!

I tried this patch in my NC 21.0.9 with user_external 2.1.0 and it seems to be working for me.

At first I seem to have made a mistake and got error 500 but I just tried again and so far so good.

I'll report at the end of the day if I get any more errors, but if you could merge this PR I'd be very grateful :)

**Comment by richarson** [comment]
So far, so good! No other issues experienced.

Can this please be merged?
Who can be pinged to review and approve the change?

**Review by violoncelloCH — APPROVED** [review]
thank you for your contribution !

**Pull request merged**